File: apps/agent/app/routes/dev.py
"""Developer simulation endpoints for testing without external provider connections."""

from fastapi import APIRouter, Depends
from app.models import SimulateInboundRequest, SimulateInboundResponse
from app.agent.runner import AgentRunner
from app.deps import get_agent_runner
from app.db.client import log_dev_outbox

router = APIRouter(prefix="/dev", tags=["dev"])


@router.post("/simulate-inbound", response_model=SimulateInboundResponse)
async def simulate_inbound(
    request: SimulateInboundRequest,
    runner: AgentRunner = Depends(get_agent_runner),
) -> SimulateInboundResponse:
    """
    Simulate an inbound message from any channel (WhatsApp, Voice, Web, etc.).

    Orchestrates session lookup, prompt assembly, mock LLM intent classification,
    and real tool dispatch via with_clinic. Logs messages to dev_outbox.
    """
    # 1. Log inbound message to dev_outbox
    log_dev_outbox(
        channel=request.channel,
        from_=request.from_,
        to="system",
        body=request.body,
        direction="inbound",
        provider="mock",
    )

    # 2. Process turn via AgentRunner
    reply, tool_intent, tool_intents, session_id = await runner.handle_inbound(
        channel=request.channel,
        sender=request.from_,
        body=request.body,
    )

    # 3. Log outbound reply to dev_outbox
    log_dev_outbox(
        channel=request.channel,
        from_="system",
        to=request.from_,
        body=reply,
        direction="outbound",
        provider="mock",
    )

    return SimulateInboundResponse(
        reply=reply,
        channel=request.channel,
        logged=True,
        session_id=session_id,
        tool_intent=tool_intent,
        tool_intents=tool_intents,
    )

File: apps/agent/src/routes/dev.ts
import { Hono } from 'hono';
import { z } from 'zod';
import { getAgentRunner } from '../deps';

export const devRoute = new Hono();

const simulateInboundSchema = z.object({
  channel: z.string().default('whatsapp'),
  from: z.string().default('+923001234500'),
  body: z.string(),
  clinic_id: z.string().optional(),
});

devRoute.post('/simulate-inbound', async (c) => {
  try {
    const rawBody = await c.req.json();
    const parsed = simulateInboundSchema.parse(rawBody);

    const runner = getAgentRunner();
    const result = await runner.handleInbound({
      channel: parsed.channel,
      sender: parsed.from,
      body: parsed.body,
      clinic_id: parsed.clinic_id,
    });

    return c.json({
      reply: result.reply,
      tool_intent: result.tool_intent,
      session_id: result.session_id,
      channel: result.channel,
      logged: result.logged,
      tool_intents: result.tool_intents,
    });
  } catch (err: unknown) {
    if (err instanceof z.ZodError) {
      return c.json({ error: 'Validation Error', details: err.errors }, 400);
    }
    const message = err instanceof Error ? err.message : 'Internal Server Error';
    return c.json({ error: message }, 500);
  }
});
